classtranscribe/FrontEnd
# fix: prevent data mangling when clicking "As a new Chapter"

## Problem
When using the ClassTranscribe FrontEnd, a user can have a chapter that contains multiple subchapters. We also offer a button to convert a subchapter into a chapter. If a subchapter isn't the last one, attempting to convert it to a chapter breaks the trailing subchapter data.

## Approach
Prevent UI from mangling this data in the same way that was used to fix `insertChapter` and `insertSubChapter`: detect if our object properties are trapped/wrapped in a single `__data__` field, and if they are then we unwrap them and use the internal data.

## How to Test
Prerequisites: signed into ct-dev as Test User

1. Create a new ePub structured as follows:
    * Chapter 1
        * Subchapter 1.1
        * Subchapter 1.2
        * Subchapter 1.3
    * Chapter 2
        * Subchapter 2.1
        * Subchapter 2.2
        * Subchapter 2.3
    * Chapter 3
        * Subchapter 3.1
        * Subchapter 3.2
        * Subchapter 3.3
2. Scroll to Subchapter 2.1 and click the "As a new Chapter" button beside it
    * You should see Subchapter 2.1 is converted to a new chapter (e.g. "Chapter 2.1")
    * You should see that the newly-formed Chapter 2.1 has Subchapters 2.2 and 2.3 as children 
    * You should see that Chapter 1 and Chapter 3 are structurally unaffected
3. Wait 5-10 seconds to ensure that changes are saved to the backend, then refresh the page
    * You should see that your changes persist across page refresh
